### PR TITLE
Reliable message delivery design proposal

### LINKED ISSUE #725: Reliable message delivery from cloud to edge
**What would you like to be added**:
Create a mechanism to make sure each message is delivered reliably to edge.

**Why is this needed**:
In offline scenarios, when there is an update in pods or device, the message is not send to edge and it is lost. So the edgeNode is not aware of the changes.

### PR DESCRIPTION
**Pull request by fisherxu**
**What type of PR is this?**

/kind design

**What this PR does / why we need it**:

Provide a reliable message delivery mechanism between cloud and edge.

**Which issue(s) this PR fixes**:

Fixes #725

**Special notes for your reviewer**:
/cc @kevin-wangzefeng @rohitsardesai83 @sids-b @kadisi @chendave @subpathdev @du2016 @cwl233

**Does this PR introduce a user-facing change?**:

```release-note
Add reliable message delivery design proposal
```

### CODE REVIEW

File: docs/proposals/reliable-message-delivery.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+  - "@fisherxu"
+  - "@kevin-wangzefeng"
+  - "@rohitsardesai83"
+approvers:
+  - "@kevin-wangzefeng"
+  - "@sids-b"
+  - "@kadisi"
+creation-date: 2019-12-23
+last-updated: 2019-12-23
+status: alpha
```
Comment (kevin-wangzefeng) [inline docs/proposals/reliable-message-delivery.md]:
```suggestion
status: provisional
```

**Comment by kevin-wangzefeng** [comment]
Comment from community meeting:
1. Add more detials(definition, behavior) about the modules in diagram
2. Clarify data structure that would be used in key-value store

File: docs/proposals/reliable-message-delivery.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+- We use K8s CRD stores the latest resourceVersion of resource that has been sent
+ successfully to edge. When cloudcore restarts or starts normally, 
+ it will check the resourceVersion to avoid sending old messages.
+ 
+- EdgeController and devicecontroller send the messages to the Cloudhub, and MessageDispatcher will send messages 
+to corresponding NodeMessageQueue according to the node name in message.
+
+- CloudHub will sequentially send data from the NodeMessageQueue to the corresponding edge node,
+ and will also store the message ID in an ACK channel. When the ACK message from the edge node received,
+ ACK channel will trigger to save the message resourceVersion to K8s as CRD, and send the next message.
```
Comment (kadisi) [inline docs/proposals/reliable-message-delivery.md]: about ack channel, I recommend using `sync.Cond` to  implementation it.

File: docs/proposals/reliable-message-delivery.md (hunk 1 above)
Comment (fisherxu) [reply docs/proposals/reliable-message-delivery.md]: Done.

File: docs/proposals/reliable-message-delivery.md (hunk 2 above)
Comment (fisherxu) [reply docs/proposals/reliable-message-delivery.md]: We have used the sync.Map now.

**Pull request merged**